app/src/ui/lib/avatar.tsx

```diff
@@ -15,74 +15,134 @@ interface IAvatarProps {
    * skipped completely if title is null
    */
   readonly title?: string | null
+
+  /**
+   * The what dimensions of avatar the component should
+   * attempt to request, defaults to 64px.
+   */
+  readonly size?: number
 }
 
 interface IAvatarState {
-  readonly dataUrl?: string
+  readonly user?: IAvatarUser
+  readonly candidates: ReadonlyArray<string>
 }
 
-/** A component for displaying a user avatar. */
-export class Avatar extends React.Component<IAvatarProps, IAvatarState> {
-  private cancelFetchingAvatar = false
-
-  public constructor(props: IAvatarProps) {
-    super(props)
-
-    this.state = { dataUrl: DefaultAvatarURL }
+const avatarEndpoint = 'https://avatars.githubusercontent.com'
+
+/**
+ * This is the person octicon from octicons v5 (which we're using at time of writing).
+ * The octicon has been tweaked to add some padding and so that it scales nicely in
+ * a square aspect ratio.
+ */
+const DefaultAvatarSymbol = new OcticonSymbol(
+  16,
+  16,
+  'M13 13.145a.844.844 0 0 1-.832.855H3.834A.846.846 0 0 1 3 13.142v-.856c0-2.257 3.333-3.429 3.333-3.429s.191-.35 0-.857c-.7-.531-.786-1.363-.833-3.429C5.644 2.503 7.056 2 8 2s2.356.502 2.5 2.571C10.453 6.637 10.367 7.47 9.667 8c-.191.506 0 .857 0 .857S13 10.03 13 12.286v.859z'
+)
+
+/**
+ * A regular expression meant to match both the legacy format GitHub.com
+ * stealth email address and the modern format (login@ vs id+login@).
+ *
+ * Yields two capture groups, the first being an optional capture of the
+ * user id and the second being the mandatory login.
+ */
+const StealthEmailRegexp = /^(?:(\d+)\+)?(.+?)@users\.noreply\.github\.com$/i
+
+/**
+ * Produces an ordered iterable of avatar urls to attempt to load for the
+ * given user.
+ */
+function* getAvatarUrlCandidates(
+  user: IAvatarUser | undefined,
+  size = 64
+): Iterable<string> {
```

Comment:
I'm not a big fan of using a generator here: since we're anyways always converting the result of `getAvatarUrlCandidates()` into an array we don't get the main benefit of generators which is laziness execution, and as a tradeoff we use a non-common language feature that may be harder to understand to some future readers (in this case I personally find this harder to understand than creating an empty array and pushing to it instead of yielding).

Is there any other reason I'm missing to use a generator here?

Reply:
> Is there any other reason I'm missing to use a generator here?

No, I just really like to yield. Iterators/generators were a huge part of the classic GitHub Desktop codebase and I was happy that ECMAScript added them to the core language.

I agree that it could look intimidating to readers of the code base though and while I think we're going to be using generators (and async generators) increasingly in the future this particular hill is not one I'm remotely willing to die on so I'll convert it to an array.